### PR TITLE
BUG: Fix Series.nlargest for integer boundary values

### LINKED ISSUE #21426: BUG: .nlargest with unsigned integers
#### Code Sample, a copy-pastable example if possible

```python
pd.Series(np.array([0, 0, 0, 100, 1000, 10000, 100], dtype='uint32')).nlargest(5) 
0        0
1        0
2        0
5    10000
4     1000

```
#### Problem description
`nlargest` favours `0` above positive values. Common to both `uint32` and `uint64` types and possibly others. 

#### Expected Output

```
5    10000
4     1000
3      100
6      100
0        0
dtype: uint32
```

#### Output of ``pd.show_versions()``
<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.2.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-327.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: en_GB.UTF-8

pandas: 0.20.3
pytest: 3.2.1
pip: 9.0.1
setuptools: 36.4.0
Cython: None
numpy: 1.13.1
scipy: 0.19.1
xarray: None
IPython: 6.1.0
sphinx: None
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: 3.4.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.8
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: 1.1.13
pymysql: 0.7.9.None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by jschendel**
- [X] closes #21426
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

Also added some similar tests for float and datetimelike dtypes to ensure that the behavior is as desired.

### CODE REVIEW

File: pandas/tests/series/test_analytics.py [hunk 1]
```diff
@@ -2028,6 +2030,33 @@ def test_n(self, n):
...
+            # use int64 bounds and +1 to min_val since true minimum is NaT
+            # (include min_val/NaT at end to maintain same expected_idxr)
+            min_val, max_val = np.iinfo('int64').min, np.iinfo('int64').max
+            vals = [min_val + 1, min_val + 2, max_val - 1, max_val, min_val]
+
+        s = Series(vals, dtype=dtype)
+        result = getattr(s, method)(3)
+        expected_idxr = [0, 1, 2] if method == 'nsmallest' else [3, 2, 1]
+        expected = s.loc[expected_idxr]
+        tm.assert_series_equal(result, expected)
```
Comment (gfyoung) [inline pandas/tests/series/test_analytics.py]: I like the thought of using parameterization, but this is a little hard to read.  I would recommend the following:

* Split this test into three tests based on which conditional it hits.
* Extract lines 2054-2058 as a separate (non-test) method, where you pass in `vals` as a parameter i.e.

~~~python
def _check_nselect(self, vals, method):
   s = Series(...)
   ...
   tm.assert_series_equal(result, expected)
~~~

**Comment by gfyoung** [comment]
> No whatsnew entry for the time being since this was tagged as 0.23.2, and no whatsnew has been created for that release yet.

@jschendel : Not a problem.  Just wait for #21433 and rebase.

**Comment by gfyoung** [comment]
#21433 has been merged.  Feel free to rebase.

File: pandas/tests/series/test_analytics.py (hunk 1 above)
Comment (jschendel) [reply pandas/tests/series/test_analytics.py]: done

**Review by gfyoung — APPROVED** [review]
Nice!

cc @jreback

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/test_analytics.py [hunk 2]
```diff
@@ -2028,6 +2032,40 @@ def test_n(self, n):
...
 
+    def _check_nselect_boundary(self, vals, dtype, method):
+        # helper function for 'test_boundary_dtype' tests
+        s = Series(vals, dtype=dtype)
+        result = getattr(s, method)(3)
+        expected_idxr = [0, 1, 2] if method == 'nsmallest' else [3, 2, 1]
+        expected = s.loc[expected_idxr]
+        tm.assert_series_equal(result, expected)
+
+    @pytest.mark.parametrize('dtype', [
```
[the remark this answers is not in the document]
Comment (gfyoung) [reply pandas/tests/series/test_analytics.py]: @jschendel brought up a good point in https://github.com/pandas-dev/pandas/pull/21456#discussion_r195266459 that we might have some intersection with this PR and #21456.  How best to work that out?

[the remark this answers is not in the document]
Comment (jschendel) [reply pandas/tests/series/test_analytics.py]: FWIW I've modified the fixtures I wrote to exactly match @gfyoung's (aside from removing float16 and fixing the `ALL_INT_DTYPES` param).  Were basically identical anyways.  I've got all my review related changes ready, so can push whenever.

**Comment by jschendel** [comment]
Addressed review comments.  Copied @gfyoung's dtype fixtures.  Did some additional cleaning of `frame/test_analytics.py.` to use one of the new fixtures I created, and to use multiple instances of `parametrize` instead of jamming everything into a single one with `product`.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jschendel

**Comment by jreback** [comment]
can you create an issue to use these fixtures project wide?

**Comment by jschendel** [comment]
> can you create an issue to use these fixtures project wide?

Created #21500 for this

**Pull request merged**